### PR TITLE
Feature/PEER-5 core values crud

### PR DESCRIPTION
**Pull request by SharyuMarwadi**

### CODE REVIEW

File: internal/pkg/apperrors/errors.go [hunk 1]
```diff
@@ -16,10 +16,29 @@ func (e CustomError) Error() string {
 
 // Custome errors with errormessage
 const (
-	BadRequest           = CustomError("Bad request")
-	InternalServer       = CustomError("Internal server error")
-	FailedToCreateDriver = CustomError("failure to create driver obj")
-	MigrationFailure     = CustomError("migrate failure")
+	InvalidId                = CustomError("Invalid id")
+	InternalServerError      = CustomError("Internal server error")
```
Comment (yama11299) [inline internal/pkg/apperrors/errors.go:20]: use smallcase, keep it uniform

File: internal/pkg/apperrors/errors.go [hunk 2]
```diff
@@ -30,10 +49,14 @@ func ErrKeyNotSet(key string) (err error) {
 // GetHTTPStatusCode returns status code according to customerror and default returns InternalServer error
 func GetHTTPStatusCode(err error) int {
 	switch err {
-	case InternalServer, FailedToCreateDriver, MigrationFailure:
+	case InternalServerError, JSONParsingErrorResp, InvalidIntranetData:
 		return http.StatusInternalServerError
-	case BadRequest:
+	case OrganizationNotFound, InvalidCoreValueData, InvalidParentValue, InvalidOrgId, GradeNotFound:
```
Comment (yama11299) [inline internal/pkg/apperrors/errors.go:54]: If invalid input is given return BadRequest.

File: internal/pkg/apperrors/errors.go [hunk 3]
```diff
@@ -30,10 +49,14 @@ func ErrKeyNotSet(key string) (err error) {
...
+		return http.StatusNotFound
+	case InvalidId, JSONParsingErrorReq, TextFieldBlank, DescFieldBlank, UniqueCoreValue, InvalidAuthToken, IntranetValidationFailed:
```
Comment (yama11299) [inline internal/pkg/apperrors/errors.go:56]: if invalid auth token, return Unauthorized.

File: internal/pkg/dto/coreValues.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package dto
+
+type UpdateQueryRequest struct {
```
Comment (yama11299) [inline internal/pkg/dto/coreValues.go:5]: Should create separate Repo and Service structs. We don't return all the data that we store in DB. Some fields are for internal use only.

File: internal/pkg/dto/coreValues.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	Name        string `json:"name" db:"name"`
+	Description string `json:"description" db:"description"`
+}
+
+type UpdateCoreValuesResp struct {
+	ID                int64  `json:"id" db:"id"`
+	Name              string `json:"name" db:"name"`
+	Description       string `json:"description" db:"description"`
+	ParentCoreValueID *int64 `json:"parent_core_value_id" db:"parent_core_value_id"`
```
Comment (yama11299) [inline internal/pkg/dto/coreValues.go:14]: can use sq.NullInt64

File: internal/repository/coreValues.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import (
+	"context"
+
+	"github.com/joshsoftware/peerly-backend/internal/pkg/dto"
+)
+
+type CoreValueStorer interface {
+	ListCoreValues(ctx context.Context) (coreValues []dto.ListCoreValuesResp, err error)
+	GetCoreValue(ctx context.Context, coreValueID int64) (coreValue dto.GetCoreValueResp, err error)
+	CreateCoreValue(ctx context.Context, userId int64, coreValue dto.CreateCoreValueReq) (resp dto.CreateCoreValueResp, err error)
```
Comment (yama11299) [inline internal/repository/coreValues.go:12]: same as above. userID from ctx or can use repo struct directly having all fields populated.

**Review by yama11299 — CHANGES_REQUESTED** [review]

File: internal/api/coreValues.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+			return
+		}
+
+		dto.SuccessRepsonse(rw, http.StatusOK, "Core value listed", coreValue)
+	})
+}
+
+func createCoreValueHandler(coreValueSvc corevalues.Service) http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		const userId int64 = 1
```
[the remark this answers is not in the document]
Comment (SharyuMarwadi) [reply internal/api/coreValues.go]: Eliminated the const

File: internal/api/coreValues.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/joshsoftware/peerly-backend/internal/pkg/dto"
+	logger "github.com/sirupsen/logrus"
+)
+
+func listCoreValuesHandler(coreValueSvc corevalues.Service) http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+
+		coreValues, err := coreValueSvc.ListCoreValues(req.Context())
+		if err != nil {
+
```
[the remark this answers is not in the document]
Comment (SharyuMarwadi) [reply internal/api/coreValues.go:20]: Logs updated

File: internal/api/coreValues.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	})
+}
+
+func createCoreValueHandler(coreValueSvc corevalues.Service) http.HandlerFunc {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		const userId int64 = 1
+		var coreValue dto.CreateCoreValueReq
+		err := json.NewDecoder(req.Body).Decode(&coreValue)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Error while decoding request data")
```
[the remark this answers is not in the document]
Comment (SharyuMarwadi) [reply internal/api/coreValues.go]: Resolved

File: internal/app/coreValues/helper.go [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	if coreValue.ParentCoreValueID != nil {
+		if !validateParentCoreValue(ctx, storer, *coreValue.ParentCoreValueID) {
+			err = apperrors.InvalidParentValue
+		}
+	}
+
+	return
+}
+
+func VarsStringToInt(inp string, label string) (result int64, err error) {
```
[the remark this answers is not in the document]
Comment (SharyuMarwadi) [reply internal/app/coreValues/helper.go]: Moved to pkg/utils/helper.go

File: internal/pkg/apperrors/errors.go (hunk 1 above)
Comment (SharyuMarwadi) [reply internal/pkg/apperrors/errors.go]: Done

File: internal/pkg/apperrors/errors.go (hunk 2 above)
Comment (SharyuMarwadi) [reply internal/pkg/apperrors/errors.go]: Done

File: internal/pkg/apperrors/errors.go (hunk 3 above)
Comment (SharyuMarwadi) [reply internal/pkg/apperrors/errors.go]: Done

File: internal/pkg/dto/coreValues.go (hunk 4 above)
Comment (SharyuMarwadi) [reply internal/pkg/dto/coreValues.go:10]: Done

File: internal/repository/coreValues.go (hunk 6 above)
Comment (SharyuMarwadi) [reply internal/repository/coreValues.go]: Done

File: internal/repository/postgresdb/coreValues.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+			"err":           err.Error(),
+			"core_value_id": coreValueID,
+		}).Error("Error while updating core value")
+		return
+	}
+
+	return
+}
+
+func (cs *coreValueStore) CheckUniqueCoreVal(ctx context.Context, text string) (isUnique bool, err error) {
```
[the remark this answers is not in the document]
Comment (SharyuMarwadi) [reply internal/repository/postgresdb/coreValues.go]: Done

File: internal/app/coreValues/service.go [hunk 12]
```diff
@@ -14,10 +16,11 @@ type service struct {
...
 type Service interface {
-	ListCoreValues(ctx context.Context) (resp []dto.ListCoreValuesResp, err error)
-	GetCoreValue(ctx context.Context, coreValueID string) (coreValue dto.GetCoreValueResp, err error)
-	CreateCoreValue(ctx context.Context, userId int64, coreValue dto.CreateCoreValueReq) (resp dto.CreateCoreValueResp, err error)
-	UpdateCoreValue(ctx context.Context, coreValueID string, coreValue dto.UpdateQueryRequest) (resp dto.UpdateCoreValuesResp, err error)
+	ListCoreValues(ctx context.Context) (resp []dto.CoreValue, err error)
+	GetCoreValue(ctx context.Context, coreValueID string) (coreValue dto.CoreValue, err error)
+	CreateCoreValue(ctx context.Context, coreValue dto.CreateCoreValueReq) (resp dto.CoreValue, err error)
+	UpdateCoreValue(ctx context.Context, coreValueID string, coreValue dto.UpdateQueryRequest) (resp dto.CoreValue, err error)
+	ValidateParentCoreValue(ctx context.Context, coreValueID int64) (ok bool)
```
Comment (yama11299) [inline internal/app/coreValues/service.go:23]: Is this getting called from other components? if no, don't export this method.

File: internal/app/coreValues/service.go [hunk 13]
```diff
@@ -26,35 +29,42 @@ func NewService(coreValuesRepo repository.CoreValueStorer) Service {
...
+func (cs *service) ListCoreValues(ctx context.Context) (resp []dto.CoreValue, err error) {
 
-	resp, err = cs.coreValuesRepo.ListCoreValues(ctx)
+	dbResp, err := cs.coreValuesRepo.ListCoreValues(ctx)
 	if err != nil {
-		logger.WithField("err", err.Error()).Error("Error while fetching data")
 		err = apperrors.InternalServerError
 	}
 
+	fmt.Println(dbResp)
```
Comment (yama11299) [inline internal/app/coreValues/service.go:39]: always remove these after testing

File: internal/app/coreValues/service.go [hunk 14]
```diff
@@ -26,35 +29,42 @@ func NewService(coreValuesRepo repository.CoreValueStorer) Service {
...
+
+	for i := 0; i < len(dbResp); i++ {
```
Comment (yama11299) [inline internal/app/coreValues/service.go:41]: can use for _, resp := range dbResp, looks better.

File: internal/app/coreValues/service.go [hunk 15]
```diff
@@ -112,13 +122,57 @@ func (cs *service) UpdateCoreValue(ctx context.Context, coreValueID string, reqD
...
+		svc := NewService(cs.coreValuesRepo)
+		if !svc.ValidateParentCoreValue(ctx, *coreValue.ParentCoreValueID) {
+			err = apperrors.InvalidParentValue
+		}
+	}
+
+	return
+}
+
+func MapCoreValueDbToService(dbStruct repository.CoreValue) (svcStruct dto.CoreValue) {
```
Comment (yama11299) [inline internal/app/coreValues/service.go:172]: can make it unexported, if no other component is directly calling corevalues' DB methods.

File: internal/app/coreValues/service.go [hunk 16]
```diff
@@ -112,13 +122,57 @@ func (cs *service) UpdateCoreValue(ctx context.Context, coreValueID string, reqD
...
+func (cs *service) Validate(ctx context.Context, coreValue dto.CreateCoreValueReq) (err error) {
+
+	if coreValue.Name == "" {
+		err = apperrors.TextFieldBlank
+	}
+	if coreValue.Description == "" {
+		err = apperrors.DescFieldBlank
+	}
+	if coreValue.ParentCoreValueID != nil {
+		svc := NewService(cs.coreValuesRepo)
```
Comment (yama11299) [inline internal/app/coreValues/service.go:163]: why NewService() when you already have coreValuesRepo in cs (*service).

File: internal/pkg/utils/helper.go [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+
+func VarsStringToInt(inp string, label string) (result int64, err error) {
+
+	if len(inp) <= 0 {
+		err = apperrors.InvalidId
+		return
+	}
+	result, err = strconv.ParseInt(inp, 10, 64)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error(fmt.Scanf("Error while parsing %s from url", label))
```
Comment (yama11299) [inline internal/pkg/utils/helper.go:19]: fix logging.

File: internal/repository/postgresdb/coreValues.go [hunk 18]
```diff
@@ -20,113 +22,129 @@ func NewCoreValueRepo(db *sqlx.DB) repository.CoreValueStorer {
...
-	// checkOrganisationQuery = `SELECT id from organizations WHERE id = $1`
-	checkUniqueCoreVal = `SELECT id from core_values WHERE name = $1`
-)
-
-func (cs *coreValueStore) ListCoreValues(ctx context.Context) (coreValues []dto.ListCoreValuesResp, err error) {
+func (cs *coreValueStore) ListCoreValues(ctx context.Context) (coreValues []repository.CoreValue, err error) {
+	queryBuilder := sq.Select("id", "name", "description", "parent_core_value_id").From("core_values")
+	listCoreValuesQuery, _, err := queryBuilder.ToSql()
+	if err != nil {
+		logger.Error(fmt.Sprintf("error in generating squirrel query, err: %s", err.Error()))
```
Comment (yama11299) [inline internal/repository/postgresdb/coreValues.go:29]: return error without logging. you can wrap error here with `fmt.Errorf("failed to create query, err:%w", err)`

File: internal/repository/postgresdb/coreValues.go [hunk 19]
```diff
@@ -20,113 +22,129 @@ func NewCoreValueRepo(db *sqlx.DB) repository.CoreValueStorer {
...
-	updateCoreValueQuery = `UPDATE core_values SET (name, description) =
-		($1, $2) where id = $3 RETURNING id, name, description, parent_core_value_id`
-
-	// checkOrganisationQuery = `SELECT id from organizations WHERE id = $1`
-	checkUniqueCoreVal = `SELECT id from core_values WHERE name = $1`
-)
-
-func (cs *coreValueStore) ListCoreValues(ctx context.Context) (coreValues []dto.ListCoreValuesResp, err error) {
+func (cs *coreValueStore) ListCoreValues(ctx context.Context) (coreValues []repository.CoreValue, err error) {
+	queryBuilder := sq.Select("id", "name", "description", "parent_core_value_id").From("core_values")
```
Comment (yama11299) [inline internal/repository/postgresdb/coreValues.go:26]: create a var corevaluesColumns []string := {"id", "name", "description", "parent_core_value_id"} and use it everywhere. 
similarly table name can be stored in coreValueStore struct or can create a const.

File: internal/app/coreValues/service.go (hunk 12 above)
Comment (SharyuMarwadi) [reply internal/app/coreValues/service.go]: Done

File: internal/app/coreValues/service.go (hunk 14 above)
Comment (SharyuMarwadi) [reply internal/app/coreValues/service.go]: Done

File: internal/app/coreValues/service.go (hunk 15 above)
Comment (SharyuMarwadi) [reply internal/app/coreValues/service.go]: Done

File: internal/app/coreValues/service.go (hunk 16 above)
Comment (SharyuMarwadi) [reply internal/app/coreValues/service.go]: Done

File: internal/pkg/utils/helper.go (hunk 17 above)
Comment (SharyuMarwadi) [reply internal/pkg/utils/helper.go:19]: Done

File: internal/repository/postgresdb/coreValues.go (hunk 19 above)
Comment (SharyuMarwadi) [reply internal/repository/postgresdb/coreValues.go]: Done

**Comment by yama11299** [comment]
approved :+1:

**Pull request merged**